etc/config/fortran.amazon.properties

```diff
@@ -0,0 +1,103 @@
+compilers=&gfortran_86:&gfortran_arm:&gfortran_ppc
```

Comment:
arm/ppc compilers don't seem to be built right now. If I recall correctly, most of the work for it is already done, no?

Reply:
Based on the experience with defining locally installed compilers (within `fortran.defaults.properties`), I assumed `compiler-explorer` would (more or less) quietly skip the ones it doesn't find and simply ignore them. But then still work with the ones it did find. So I assumed once all compilers do actually pop up it would make them all available.

From a quick look it seems the `install-compilers.sh` only grabs tar-balls to unpack them. Poking around a little more I see files that seem to hold the config (`arm-7.1.0.config
`) but say that they are automatically generated and not to be edited? I'll search around little more tomorrow to see if I can find where/how these are generated.

Reply:
Yes, you are right. Unreachable compilers are safely ignored. What's different about this config is that it directly affects the live site, so we are a bit more strict with it than with the rest of the configuration files.

Regarding the building of compilers, I'll set it up and let you know once things are properly set up.


Thanks once again for your patience :)

Reply:
A live site will be treated differently, that's right. I also don't think I've even scratched the surface of the project, so absolutely your call. Don't want to appear pushy either.

Let me know if there's anything I can do to help in this case.

Reply:
So, the x86 compilers are all built from source and packed into tarballs by https://github.com/mattgodbolt/compiler-explorer-image/blob/master/gcc/build/build.sh

The cross-compilers are sometimes built using crosstool-ng, and sometimes downloaded from vendor websites. The ARM ones in particular are a mix of both types. As cross-compilation is such a pain (even with crosstool-ng), I've not put much effort into getting non-C++ compilers built, and of course with the vendor tarballs we get whatever they decide to ship.

I'd be very grateful for help in getting the cross-compilation working a lot better!

Reply:
I was poking around the other repo a bit. I found a `build.sh` in `gcc-cross/build` that is used to build the cross-compilers via `crosstool-ng`. Changing its configs to include Fortran support seems easy. I've tested it for 2 versions of GCC so far and it appeared to work ok.
I can continue to check for what versions of GCC it correctly builds gfortran.

However:
1) Running the `Dockerfile` I get an error when the 'old' version of `crosstool-ng` is being built (the build aborts). I'll check if the new version can in fact be used to build the older compilers as well.
2) I grep'ed  through the repo and can't find any script actually running `gcc-cross/build/build.sh`. E.g. `admin-daily-builds.sh` appears to run `gcc/build/build.sh` routinely, but I find no such reference for the cross-compilers. I have no clue how you generated the cross-compilers, is it possible that you ran it manually at some point?
3) Our admins tell me that the newer ARM compilers need licenses (the vendor download). That's no business I want to get into. From my point of view only you guys can tell where you do or don't have licenses.
4) My impression of `crosstool-ng` is a little mixed. Running its menuconfig I only get a choice of 3 GCC versions - not even all the ones you seem to have configs for. So I'm skeptical what will and won't work, e.g. I would have attempted to include GCC8.

Reply:
Hi @dbroemmel indeed! At the moment the cross-compilation build is - as you've spotted - broken. It died through an upgrade to Ubunu 18.04.

The cross compilers are not built as part of any automatic process. We don't build cross-compilers daily, only when needed.

Newer ARM compilers need licenses?! That's news to me :( I would have thought anything based on GPL'd GCC would have to work without a license.

My thoughts on crosstool-ng are similar. That's why there are several different versions of crosstool-ng built into the docker image, each supports a (different) set of versions of compilers. They're not great. Any ideas on how to improve the situation gratefully accepted.

Cross-compilers are not compilers I use myself, so I'm always looking for folks who know more than this about me to help!

Reply:
Hi @mattgodbolt. I have been fiddling with it for another 2-3 days, mostly trying to ditch `crosstool-ng`. Some luck, but not consistent and clean enough to commit or document anything. I've figured out a few bits I think are necessary, but still have no clue about the intricate details. I was thinking about taking shortcuts using the multiarch capabilities of the Ubuntu 18.04 image, but I'm not sure if this would suit your needs.

My last attempt was to model it along the files in `gcc/build` (assuming your patches and configure tweaks would help as a start), but don't even seem to succeed building plain compilers with those, let along cross-compilers. When I was able to build a cross-compiler, it was the 7.3.0 version, no luck with earlier ones.

I'm not sure about ARM compilers (and didn't check the web), but was told the newer ones build on llvm and need a license (so this likely doesn't include Fortran).

Do the missing cross-compilers really have to block the Fortran side of things for x86?